docs/adr/006_cups.md

```diff
@@ -0,0 +1,296 @@
+# 006. CUPS Protocol Implementation - Credential management
+
+**Feature**: [#391](https://github.com/Azure/iotedge-lorawan-starterkit/issues/391)
+
+**Date**: 19 November 2021
+
+**Author**: Daniele Antonio Maggio
+
+**Status**: Proposed
+
+__________
+
+## Overview
+
+LoRa Basics™ Station defines a CUPS protocol for providing updated LNS/CUPS credentials and generic update data binary to any Client connecting to a CUPS Server.
+
+More information on the protocol can be found [here](https://doc.sm.tc/station/cupsproto.html).
+
+## In-scope
+
+This document focuses on:
+
+- Defining a flow diagram for common credential management scenarios:
+  - First connection
+  - Certificate rotation process
+- Defining the changes needed in IoT Hub for handling Basics Station specific information
+- Defining the changes needed in 'Facade' Azure Function for handling updated credential retrieval
+- Defining the changes needed in LoRaWan Network Server for handling above mentioned flows
+- Defining the changes needed in LoRa Device Provisioning CLI for handling creation/update of concentrators with certificate support
+
+## Out-of-scope
+
+Out of scope for this document is:
+
+- Defining a flow diagram for firmware update process
+- Defining any change needed in IoT Hub / Azure Function / Network Server for handling firmware updates
+
+## Authentication mechanisms with LoRa Basics™ Station
+
+### Context
+
+To better understand how the CUPS Protocol should be implemented, it is needed to understand how Basics Station is handling authentication.
+
+LoRa Basics™ Station supports four different authentication modes:
+
+- No Authentication
+- TLS Server Authentication
+- TLS Server and Client Authentication
+- TLS Server Authentication and Client Token
+
+An authentication mode can be set by configuring some files in the device where Basics Station is going to be executed.
+
+Basics Station can either use two different sets of credentials for achieving client authentication to the CUPS endpoint and to the LNS endpoint or re-use the same set of credentials for both.
+
+The aim of this document section is to take and record a decision for:
+
+- Which authentication mode needs to be used for connecting to CUPS endpoint
+- Which authentication mode needs to be used for connecting to LNS endpoint
+- Where are the certificates stored and how to retrieve those?
+
+### Decision(s)
+
+Out of the four authentication modes we should aim for "TLS Server and Client Authentication" for both CUPS and LNS endpoints.
```

Comment:
[Brainstorm] Not sure it's a good idea, but Like to discuss

What if we used the client token auth instead of certificate for client. We could generate a client token with the station's private key in IoT Hub (signature-string). We would then verify the token by simply trying to connect to the IoT Hub.

The advantage I see is No need to provision a client certificate on the LNS for validation as the authorization procedure is offloaded to the iot hub.

Reply:
Agree that there's no need to provision a client certificate, but there would be need to create a tool generating such token and anyways you would need to provision the token to the device, no?

Reply:
As discussed orally, the above idea would provide an interresting simplification of required implementation. However. if we want to permit rotation of the tc.trust file, the gain would be very low.